### ISSUE TITLE
Networking Thread Safety

### ISSUE THREAD

**Issue by blay09** [zerofall/EZStorage2#30]
Hi, unless I missed something when reading the code it looks like your network packet handling is not correctly adapted for Minecraft 1.8+ and is likely to cause obscure thread-related crashes that are hard to track down (and often make it look like some other mod is causing them).

> As of Minecraft 1.8 packets are by default handled on the network thread.
> 
> That means that your `IMessageHandler` can not interact with most game objects directly. [...] Minecraft provides a convenient way to make your code execute on the main thread instead using `IThreadListener.addScheduledTask`.
> 
> The way to obtain an `IThreadListener` is using either the `Minecraft` instance (client side) or a `WorldServer` instance (server side).

Source: http://mcforge.readthedocs.io/en/latest/networking/simpleimpl/

Basically, your packet handling code needs to be wrapped in a Runnable and scheduled for execution on the main thread via the method mentioned above to prevent weird, random issues.

**Comment by sblectric** [comment]
Yeah, I've known about this. Haven't had any crashes yet, but I will eventually make the fix unless someone makes a PR on it.

**Comment by blay09** [comment]
The crashes are most frequent in modpacks where there's lots of stuff going on, and are often reported on the wrong mods (see https://github.com/aidancbrady/Mekanism/issues/3361 which was reported to JEI, InventoryTweaks, MCMultipart and BMC, and only tracked back to Mekanism due to the user noticing the connection).

**Comment by sblectric** [comment]
Good to know, thanks.

**Comment by kreezxil** [comment]
This might be related to this. I've been getting odd dedicated server crashes, more like hangs, because there appear to be no crash reports in regard to it. I'll log on and see that my server has stalled midstream, I'll get on my console and notice that it is stuck, not accepting commands, so I'm forced to kill it in the most vicious way possible on Linux which is to send it SIGKILL signal.

I was digging about in all of my log files including the compressed ones and found this from the other day, hopefully this small excerpt is related to this and we can all put this baby to bed.

```[21:57:35] [Server thread/FATAL]: Error executing task
java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:1.8.0_101]
        at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[?:1.8.0_101]
        at net.minecraft.util.Util.func_181617_a(SourceFile:46) [h.class:?]
        at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:668) [MinecraftServer.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:387) [ld.class:?]
        at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:613) [MinecraftServer.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:471) [MinecraftServer.class:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
Caused by: java.lang.NullPointerException
        at com.zerofall.ezstorage.util.EZInventory.stacksEqualOreDict(EZInventory.java:223) ~[EZInventory.class:?]
        at com.zerofall.ezstorage.util.EZInventory.getItems(EZInventory.java:179) ~[EZInventory.class:?]
        at com.zerofall.ezstorage.gui.server.ContainerStorageCoreCrafting.tryToPopulateCraftingGrid(ContainerStorageCoreCrafting.java:140) ~[ContainerStorageCoreCrafting.class:?]
        at com.zerofall.ezstorage.gui.server.ContainerStorageCoreCrafting.func_82846_b(ContainerStorageCoreCrafting.java:84) ~[ContainerStorageCoreCrafting.class:?]
        at net.minecraft.inventory.Container.func_184996_a(Container.java:258) ~[abd.class:?]
        at com.zerofall.ezstorage.gui.server.ContainerStorageCore.func_184996_a(ContainerStorageCore.java:93) ~[ContainerStorageCore.class:?]
        at com.zerofall.ezstorage.gui.server.ContainerStorageCoreCrafting.func_184996_a(ContainerStorageCoreCrafting.java:116) ~[ContainerStorageCoreCrafting.class:?]
        at net.minecraft.network.NetHandlerPlayServer.func_147351_a(NetHandlerPlayServer.java:1112) ~[me.class:?]
        at net.minecraft.network.play.client.CPacketClickWindow.func_148833_a(SourceFile:33) ~[ir.class:?]
        at net.minecraft.network.play.client.CPacketClickWindow.func_148833_a(SourceFile:10) ~[ir.class:?]
        at net.minecraft.network.PacketThreadUtil$1.run(SourceFile:13) ~[fl$1.class:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_101]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_101]
        at net.minecraft.util.Util.func_181617_a(SourceFile:45) ~[h.class:?]
        ... 5 more
```

**Comment by sblectric** [comment]
Maybe. I'll do the fix soon and we shall see.

**Comment by sblectric** [comment]
Will be fixed in 2.2.3.
